helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -1310,26 +1318,30 @@ public void run() throws Exception {
   }
 
   private void fireChildChangedEvents(final String path, Set<IZkChildListener> childListeners) {
+    //TODO: Use event type to check if path exists and save network request
+    boolean pathExists = exists(path);
     try {
-      final ZkPathStatRecord pathStatRecord = new ZkPathStatRecord(path);
       for (final IZkChildListener listener : childListeners) {
         _eventThread.send(new ZkEvent("Children of " + path + " changed sent to " + listener) {
           @Override
           public void run() throws Exception {
-            // Reinstall watch before listener callbacks to check the znode status
-            if (!pathStatRecord.pathChecked()) {
-              pathStatRecord.recordPathStat(getStat(path, hasListeners(path)),
-                  OptionalLong.empty());
-            }
-            List<String> children = null;
-            if (pathStatRecord.pathExists()) {
+            List<String> children = new ArrayList<>();
```

Comment:
Do you change this list initialization from null to an instance for the later for loop? But do you expect the listener will be removed in any case in `handleChildChange`? From the code in `handleChildChange`, if `children` is not null, the listener will not be removed.

Reply:
Thanks for mentioning it! You're talking about CallbackHandler's implementation, right? I just checked you're right, that CallbackHandler does rely on the NULL to remove the listener.
To me, it's very odd cause it obviously violates the decoupling principle while 2 different classes are so tightly correlated, also it doesn't make sense to pass a NULL variable. 

I will put a TODO and leave the current code as it is. But still, need to add the child exists watcher part.